nvmeof-recoverer:  reassign devices from faulted OSDs to new nodes

cheolho-kang opened the pull request:
<!-- Thank you for contributing to Rook! -->

<!-- STEPS TO FOLLOW:
  1. Add a description of the changes (frequently the same as the commit description)
  2. Enter the issue number next to "Resolves #" below (if there is no tracking issue resolved, **remove that section**)
  3. Review our Contributing documentation at https://rook.io/docs/rook/latest/Contributing/development-flow/
  4. Follow the steps in the checklist below, starting with the **Commit Message Formatting**.
-->

<!-- Uncomment this section with the issue number if an issue is being resolved
**Issue resolved by this Pull Request:**
Resolves #
--->

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

cheolho-kang committed:
nvmeof-recoverer: add osdIDfield into nvmeofstorage CR

- modified types.go to add 'osdID' to the FabricDevice struct
- modified reconcile function to save osdID into the nvmeofstorage CR after updating the CRUSH map
- 'deploy/' and 'Documentation/' directories are auto-generated code

signed-off-by: cheolho.kang <cheolho.kang@samsung.com>

cheolho-kang committed:
nvmeof-recoverer: delete faulted OSD and disconnect associated fabric device

- added code to delete an OSD deployment in `CrashLoopBackOff` state
- added functionality to disconnect the fabric device previously used by the faulted OSD

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -217,8 +218,35 @@ func (r *ReconcileNvmeOfStorage) Reconcile(context context.Context, request reco
+		}
+
+		// Connect the device to the new host
+		newDevice, err := r.clustermanager.RunNvmeoFConnectionJob(cm.NvmeofConnect, nextHostName,
```
HostInfo 라는 구조에 뒤에 3개 arguments 가 추상화 될 수 있을 것 같습니다.

cheolho-kang replied:
넵, 앞으로 함수들을 만들때 argument를 최소화시키는 것을 생각하며 작성해보겠습니다. 수정하였습니다!

작성하다보니 disconnect일때는 subnqn만 필요해서, 불필요한 변수들은 모두 제거했습니다. 참고부탁드립니다

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -237,6 +238,26 @@ func (r *ReconcileNvmeOfStorage) Reconcile(context context.Context, request reco
+		logger.Debugf("successfully deleted the OSD deployment. Name: %q", deploymentName)
+
+		// Disconnect the device used by this OSD
+		_, err = r.clustermanager.RunNvmeoFConnectionJob(cm.NvmeofDisconnect, nextHostName,
```
명시적으로 disconnect 라고 function 을 call 하는게 argument 1 을 줄일 수 있을 것 같습니다.

cheolho-kang replied:
넵! 이부분이 함수 이름도 그렇고 계속 애매했는데 훨씬 좋은 것 같습니다. 수정하였습니다!

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -217,8 +218,39 @@ func (r *ReconcileNvmeOfStorage) Reconcile(context context.Context, request reco
+		}
+
+		// Connect the device to the new host
+		hostInfo := cm.HostInfo{
```
spec에 아예 넣는건 어떤가요?

cheolho-kang replied:
말씀하신 것처럼 spec에 넣으려면 FabricDevice 내용 중 HostInfo에 해당하는 변수들을 추출해서 따로 선언해보고자 하는데요.
FabricDevice들의 공통 정보인 NvmeOfStorageSpec.IP가 상위 구조체에 있어서 쉽진 않아보입니다..
함수 argument로 ip, hostinfo 를 나눠서 넣는다면 아래와같이 spec에 정의해서 사용 가능할 것 같습니다.

![image](https://github.com/myoungwon/rook/assets/13426789/538f1c61-a371-4fbf-a52b-c711ef75069e)

myoungwon replied:
넵 어차피 필요한 내용인것 같습니다

cheolho-kang replied:
수정하였습니다.
`r.nvmeOfStorage.Spec` 를 통째로 넣고싶었는데, 오히려 arugment가 많아져서 ip와, `cephv1.FabricDevice`를 넣도록 수정하였습니다.
왜냐하면 spec을 통째로 넣을 경우 `ConnectNvmeoF()` 코드 내에서 spec 내 device들 중 faulted OSD의 hostname과 device name으로  `cephv1.FabricDevice`를 찾아야했습니다..
`getAttachedDeviceInfo()`를 따로 만들어서 connect, disconnect할 device info만 추출하도록 하고, `ConnectNvmeoF()` 함수 호출시 argument는 ip와 fabric device info를 넣었습니다

cheolho-kang replied:
의견 주신 내용 참고하여 다시 수정하였습니다.
- ConnectNvmeoF 함수 내에서 target device 정보들을 찾고 job 호출하도록 수정

한가지 고민사항이 있습니다.
- disconnect와 달리 connect에서는 CR을 찾는 과정이 포함되어있어서 함수 성향이 좀 달라진 것 같습니다.

건의사항
- `ReassignFaultedOSDDevice()`을 만들어서 CR을 찾는 것, `GetNextAttachableHost()`, 그리고 `ConnectNvmeoF()`를 묶는건 어떨지 의견 여쭤봅니다

cheolho-kang replied:
reverse map을 만들고,
connect 함수 호출시 nqn만 arugment로 넣어서 reversemap을 이용해 추가 정보들을 내부에서 삽입하도록 수정하였습니다.

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/clustermanager/handler.go:
```diff
@@ -38,3 +115,91 @@ func (cm *ClusterManager) GetNextAttachableHost(currentHost string) string {
+		},
+	}
+
+	err := k8sutil.RunReplaceableJob(cm.opManagerContext, cm.context.Clientset, job, false)
```
disconnect 할때도 필요한가요?
RunReplicateableJob 의 역할이 무엇 일까요?

cheolho-kang replied:
K8S에서 다른 노드에게 `일회성 작업`을 수행시키기 위한 job 실행 함수라 필요합니다. 
그리고 특별히 해당 함수를 사용한 이유는 connect/disconnect job은 compute node fail시 계속 사용될거라서 같은 이름으로 job을 재사용하기 위함입니다

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -217,8 +219,58 @@ func (r *ReconcileNvmeOfStorage) Reconcile(context context.Context, request reco
+			}
+		}
+		// Retrieve the fabric device info for next host connection
+		subNQN := ""
```
`// Delete the OSD ...`
밑으로 내려도 될 것 같습니다.

cheolho-kang replied:
넵! 수정하였습니다

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -217,8 +218,38 @@ func (r *ReconcileNvmeOfStorage) Reconcile(context context.Context, request reco
 
-		// TODO: Add create and run job for nvme-of device switch to the next host
-		// Placeholder for the job creation and execution
+		// Delete the OSD deployment that is in CrashLoopBackOff
```
add cleanup()

cheolho-kang replied:
넵, cleanup으로 묶었습니다

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -197,39 +198,76 @@ func (r *ReconcileNvmeOfStorage) Reconcile(context context.Context, request reco
 
 		// Get the next attachable hostname for the OSD
-		nextHostName := r.clustermanager.GetNextAttachableHost(attachedNode)
+		nextHostName := r.clustermanager.GetNextAttachableHost(deviceInfo.AttachedNode)
```
reassign 으로 추상화

cheolho-kang replied:
reassign 으로 추상화 해보다가 한가지 걸리는점이 생겼습니다.
이 다음 PR인 cephCluster CR을 업데이트 과정에서
새로 할당한 node가 어떤 것인지 표기를 해야해서 nextHostName이 필요한데,

만약 reassign으로 추상화하게 되면, nextHostName을 어떻게 가져와야할지 고민되는 것 같습니다.

1. resassing 으로 추상화한 함수가 return 값으로, nextHostName과 새로 할당한 deviceName을 같이 return하도록 함.
2. 이번에 추가한 reversemap을 이용해서 clustermanager에 getter 함수를 만들어서 가져오도록 함.

의견주시면 반영하여 수정하겠습니다

myoungwon replied:
deviceInfo 만 써도 될듯 합니다.

cheolho-kang replied:
말씀해주신 내용 참고해봤을 때

deviceInfo를 포인터로 넣고, 새로운 node에 device connect가 성공했을 때 deviceInfo를 함수 내부에서 직접 수정하는 방식으로하면 더 깔끔할 것 같아 수정해봤습니다. 확인부탁드려요.

cheolho-kang replied:
포인터를 사용하지 않도록 다시 수정하였습니다.

`reassignFaultedOSDDevice()`는 추상화가 목적이라 이함수 내에서는 포인터 사용을 유지시켜놨는데, 이부분도 수정이 필요해보이신다면 의견 부탁드립니다.

myoungwon replied:
output를 리턴받아서 대입하는 건 어떨까요?

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -197,39 +198,76 @@ func (r *ReconcileNvmeOfStorage) Reconcile(context context.Context, request reco
+		logger.Debugf("Pod %q is going be transferred from %s to %s", request.Name, deviceInfo.AttachedNode, nextHostName)
+
+		// Connect the device to the new host
+		newDevice, err := r.clustermanager.ConnectNvmeoF(nextHostName, r.nvmeOfStorage, osdID)
```
논의한대로 osdID 삭제하여 connectNvmeoF 재구성

cheolho-kang replied:
reversemap 없이 해보려 했으나, 결국 역으로 ip를 찾는 과정이 발생하여 의견주신것처럼 reversemap을 사용하는게 가장 명료할 것 같았습니다 ㅎㅎ.. 의견주셔서 감사합니다

cheolho-kang committed:
nvmeof-recoverer: add reverse map for NQN to the endpoint

cheolho-kang committed:
nvmeof-recoverer: refactor deviceInfo retrieval to use only OSD ID

Signed-off-by: cheolho.kang <cheolho.kang@samsung.com>

cheolho-kang committed:
nvmeof-recoverer: add job to manage nvme-of device connection

- implemented ConnectNvmeoF() to connect nvme-oF devices across nodes
- included context and opManagerContext into the ClusterManager for job execution

Signed-off-by: cheolho.kang <cheolho.kang@samsung.com>

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -197,33 +198,59 @@ func (r *ReconcileNvmeOfStorage) Reconcile(context context.Context, request reco
+		logger.Debugf("Pod %q is going be transferred from %s to %s", request.Name, deviceInfo.AttachedNode, nextHostName)
+
+		// Connect the device to the new host
+		newDevice, err := r.clustermanager.ConnectNvmeoF(nextHostName, deviceInfo.SubNQN)
```
deviceInfo 만 argument로 넣어도 될듯

cheolho-kang replied:
넵 수정하였습니다

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -232,6 +259,28 @@ func (r *ReconcileNvmeOfStorage) getPods(context context.Context, namespace stri
+	}
+
+	// Disconnect the device used by this OSD
+	_, err = r.clustermanager.DisconnectNvmeoF(deviceInfo)
```
함수의 뜻이 모호함. DisconnectDevice?

cheolho-kang replied:
좋은 의견 감사합니다. Connect func과 함께 고려했을때 통일성있게 정의할 수 있을 것 같아 좋아보입니다.
DisconnectDeviceFromHost는 혹시 너무 길까요?

myoungwon replied:
조금 너무 길어 보입니다.ㅎㅎ

cheolho-kang replied:
넵
`DisconnectOSDDevice()`
`ConnectOSDDeviceToHost()`

으로 수정하였습니다

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -197,33 +198,59 @@ func (r *ReconcileNvmeOfStorage) Reconcile(context context.Context, request reco
-		logger.Debugf("Pod %q is going be transferred from %s to %s", request.Name, attachedNode, nextHostName)
+		logger.Debugf("Pod %q is going be transferred from %s to %s", request.Name, deviceInfo.AttachedNode, nextHostName)
+
+		// Connect the device to the new host
```
함수의 뜻이 모호함 ConnectNvmeof?

cheolho-kang replied:
ConnectDeviceToHost 는 혹시 어떤가요?

myoungwon replied:
ConnectOSDDeviceToHost ?

cheolho-kang replied:
넵!
